### PR TITLE
Fixed 'Qualify member access' analyzer for object initializer

### LINKED ISSUE #22776: Incorrect "Add 'this.'" diagnostic for class initializers
This is reported in a few Developer Community feedback items, I will link them as they're processed.
-  https://developercommunity.visualstudio.com/content/problem/131432/add-this-or-me-qualification-for-class-initializer.html
- https://developercommunity.visualstudio.com/content/problem/129705/incorrect-ide0009-in-object-initializer.html
- https://developercommunity.visualstudio.com/content/problem/161932/ide0009-add-this-or-me-qualification-incorrectly-r.html

I have not been able to reproduce this locally using the settings in the Tools | Options. Maybe it requires the editorconfig? That doesn't make much sense but I haven't been able to come up with anything else.

### LINKED ISSUE #23016: Add this or me qualification rule triggered incorrectly in return statement
Have the following code:

            return new UserSession
            {
                UserId = subject
            };

Where UserSession is a simple class with property UserId. I'm getting green squiggles under the UserId in that statement, suggesting I add this or me qualification.

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/126539/add-this-or-me-qualification-rule-triggered-incorr.html
VSTS ticketId: 505612_
_These are the original issue comments:_

David Poeschl [MSFT] on ‎10‎/‎18‎/‎2017, 00:37 PM (17 days ago): <p>Thank you for your feedback! I've so far been unable to reproduce this problem. I tried with the following code:</p><p><br></p><p>class C</p><p>\{</p><p>  static UserSession M()</p><p>  \{</p><p>    object subject = null;</p><p>    return new UserSession</p><p>    \{</p><p>      UserId = subject</p><p>    };</p><p>  }</p><p>}</p><p><br></p><p>internal class UserSession</p><p>\{</p><p>  public object UserId \{ get; set; }</p><p>}</p><p><br></p><p>And under &quot;Text Editor | C# | Code Style | General | 'this.' preferences&quot; I've set everything to &quot;Prefer 'this.'&quot; with &quot;Warning&quot; severity. Am I missing a step that you can tell? Thanks!</p>

_These are the original issue solutions:_
(no solutions)

### PR DESCRIPTION
**Pull request by zaytsev-victor**
Fixes #22776
Fixes #23016

Regression from 15.4

VSO bug : [link](https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/529870)

### CODE REVIEW

**Comment by sharwell** [comment]
@jinujoseph @MattGertz This is a new regression in 15.5. Can we get a bar check on it?

**Comment by MattGertz** [comment]
@sharwell Yes, I would like to take this to 15.5 -- please create an appropriately decorated 15.5 VSO bug & tag me there.

**Comment by sharwell** [comment]
### Customer scenario

User applies a Fix All operation for "Use Object Initializer" while also having a preference for qualifying member access with `this.` or `Me.`. The resulting code fix tries to qualify the left hand side of the initializer, which is a syntax error.

### Bugs this fixes

Fixes #22776
Fixes #23016
Fixes #23381 
Fixes [DevDiv 529870](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/529870)

### Workarounds, if any

Disable the Use Object Initializer analyzer, and/or disable the Qualify Member Access analyzer.

### Risk

Low. The test suite is substantially expanded by this as well.

### Performance impact

*Not yet evaluated*

### Is this a regression from a previous update?

Yes, this is a new regression in 15.5 (first appeared in 15.5 Preview 1).

### Root cause analysis

Failed to perform churn (differential) testing of diagnostics reported by our analyzers to detect changes in large code bases. The upcoming AnalyzerRunner tool should enable testing these scenarios on a more regular basis.

### How was the bug found?

Customer reported (internal and external).

### Test documentation updated?

No.

File: src/Features/VisualBasic/Portable/Diagnostics/Analyzers/VisualBasicQualifyMemberAccessDiagnosticAnalyzer.vb [hunk 1]
```diff
@@ -18,7 +18,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.QualifyMemberAccess
 
         Protected Overrides Function CanMemberAccessBeQualified(containingSymbol As ISymbol, node As SyntaxNode) As Boolean
             ' If the member is already qualified with `MyBase.`, or `MyClass.`, it cannot be further qualified.
```
Comment (sharwell) [inline src/Features/VisualBasic/Portable/Diagnostics/Analyzers/VisualBasicQualifyMemberAccessDiagnosticAnalyzer.vb]: :bulb: Consider updating this documentation to describe the new scenario.

File: src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -18,7 +18,9 @@ protected override bool IsAlreadyQualifiedMemberAccess(SyntaxNode node)
 
         // If the member is already qualified with `base.`, it cannot be further qualified.
```
Comment (sharwell) [inline src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: :bulb: Consider updating this documentation to describe the new scenario.

File: src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -16,9 +16,13 @@ protected override string GetLanguageName()
...
 
-        // If the member is already qualified with `base.`, it cannot be further qualified.
+        // If the member is already qualified with `base.`,
+        // or member is in object initialization context,
+        // or member in property or field initialization, it cannot be qualified.
         protected override bool CanMemberAccessBeQualified(ISymbol containingSymbol, SyntaxNode node)
-            => !(node.IsKind(SyntaxKind.BaseExpression) || IsInPropertyOrFieldInitialization(containingSymbol, node));
+            => !(node.IsKind(SyntaxKind.BaseExpression) ||
+                node.Parent.Parent.IsKind(SyntaxKind.ObjectInitializerExpression) ||
+                IsInPropertyOrFieldInitialization(containingSymbol, node));
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: any concern about a similar issue in tuples?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/QualifyMemberAccess/CSharpQualifyMemberAccessDiagnosticAnalyzer.cs]: nit: personally, i would prefer:  ```!... && !... && !... over !(... || ... | ...).```

**Comment by natidea** [comment]
Approved to merge via https://devdiv.visualstudio.com/DevDiv/_workitems/edit/529870

**Comment by bignorm64** [comment]
@sharwell IDE009 is still incorrectly reported for a collection initializer that 'new's up its elements.

E.g. Given

```C#
class Cat
        {
            public Cat()
            {
            }

            public Cat(string name, int age)
            {
                this.Name = name;
                this.Age = age;
            }

            public int Age { get; set; }

            public string Name { get; set; }
        }
```
The array initializer below does not report any IDE009 warnings but the list initializer does.

```C#
            // Array initializer is fine
            Cat[] catArray = new Cat[]
            {
                new Cat("Whiskers", 5),
                new Cat() { Name = "Sylvester", Age = 8},
            };

            // IEnumerable initializer reports IDE009
            List<Cat> cats = new List<Cat>
            {
                new Cat("Whiskers", 5), // IDE009 here
                new Cat() { Name = "Sylvester", Age = 8}, // and IDE009 here

            };

```

**Pull request merged**